### PR TITLE
Sbratton/osgi connect loader

### PR DESCRIPTION
**Pull request by sebratton**

### CODE REVIEW

File: atomos.framework/src/main/java/org/atomos/framework/modules/ModuleConnectLoader.java [hunk 1]
```diff
@@ -67,22 +86,47 @@ void initEdges(Module module, Configuration loaderConfig, List<ModuleLayer> pare
      */
     @Override
     protected URL findResource(String moduleName, String name) throws IOException {
-    	// TODO
-    	// The moduleName is the module which is trying to find the resource.
-    	// If the module name is not equal to this loaders module name, return null
-    	
-    	// Otherwise use the reader to find the named resource; should protect with doPriv
-    	return null;
+        URL resource=null;
```
Comment (tjwatson) [inline atomos.framework/src/main/java/org/atomos/framework/modules/ModuleConnectLoader.java]: Why is this commented out?

File: atomos.framework/src/main/java/org/atomos/framework/modules/ModuleConnectLoader.java [hunk 2]
```diff
@@ -67,22 +86,47 @@ void initEdges(Module module, Configuration loaderConfig, List<ModuleLayer> pare
...
 
     @Override
     public URL findResource(String name) {
-        // get the package name of the resource
-    	// check if the ModuleDescriptor for the module reference contains the package name in its packages()
-    	// If not return null
-    	// otherwise use findResource(String, String) with the assumption that this module is driving the find
-    	// If the resource does not end in ".class" then check if the package is open for reflection.
-    	// non-class resources should be returned by this method if the package is open
+//        // get the package name of the resource
```
Comment (tjwatson) [inline atomos.framework/src/main/java/org/atomos/framework/modules/ModuleConnectLoader.java]: Seems this is still work to be done

File: atomos.framework/src/main/java/org/atomos/framework/modules/ModuleConnectLoader.java [hunk 3]
```diff
@@ -116,30 +160,86 @@ public URL getResource(String name) {
...
+            throw new ClassNotFoundException("Could not find class: "+className);
+        }
+    	return cls;
     }
 
     @Override
     protected Class<?> findClass(String moduleName, String className) {
-        // do the same as findClass(String className) except return null instead of exception on not found
-    	return null;
+//        Class<?> cls=null;
```
Comment (tjwatson) [inline atomos.framework/src/main/java/org/atomos/framework/modules/ModuleConnectLoader.java]: Why is this commented out code still here?

File: atomos.framework/src/main/java/org/atomos/framework/modules/ModuleConnectLoader.java [hunk 4]
```diff
@@ -29,32 +37,43 @@
...
     private final AtomicReference<Module> module = new AtomicReference<>();
 
+    private HashMap<String, ClassLoader> edges = new HashMap<String, ClassLoader>();
+
     public ModuleConnectLoader(ResolvedModule resolvedModule, AtomosRuntimeModules atomosRuntimeModules) throws IOException
     {
         super("ModuleConnectLoader-" + resolvedModule.name(), null);
 
         this.resolvedModule = resolvedModule;
+        //TODO remove reference and reader? Or Reader needs to be closed?
```
[the remark this answers is not in the document]
Comment (sebratton) [reply atomos.framework/src/main/java/org/atomos/framework/modules/ModuleConnectLoader.java]: partially resolved and left a new TODO in the code

File: atomos.framework/src/main/java/org/atomos/framework/base/AtomosRuntimeBase.java [hunk 5]
```diff
@@ -108,6 +108,13 @@ public static AtomosRuntimeBase newAtomosRuntime() {
 		}
 	}
 
+	/** explicit call from AtomosRuntimBundle to allow resources to be freed.
```
Comment (tjwatson) [inline atomos.framework/src/main/java/org/atomos/framework/base/AtomosRuntimeBase.java]: I'm not clear what will call this.  I think we need a close method on `ModuleConnectLoader` that would get called by `AtomosRuntimeModules.removeLayer` method

File: atomos.framework/src/main/java/org/atomos/framework/modules/ModuleConnectLoader.java [hunk 6]
```diff
@@ -67,36 +140,82 @@ void initEdges(Module module, Configuration loaderConfig, List<ModuleLayer> pare
...
     @Override
     public URL findResource(String name) {
-        // get the package name of the resource
-    	// check if the ModuleDescriptor for the module reference contains the package name in its packages()
-    	// If not return null
-    	// otherwise use findResource(String, String) with the assumption that this module is driving the find
-    	// If the resource does not end in ".class" then check if the package is open for reflection.
-    	// non-class resources should be returned by this method if the package is open
-    	return null;
+    	URL retVal = null;
```
Comment (tjwatson) [inline atomos.framework/src/main/java/org/atomos/framework/modules/ModuleConnectLoader.java]: My original comments mislead you in the implementation here:

If the package name is contained in the java.lang.module.ModuleDescriptor.packages() set then
- check the findResource(String, String) method using the module name associated with the loader
- if a resource is found then do a check for ".class" extension or if the package is open and if so then
- return the resource

If the package name is NOT contained in the java.lang.module.ModuleDescriptor.packages() set then
- simply return the result of findResource(String, String) method using the module name associated with the loader

**Pull request merged**